go/consensus/tendermint/apps/beacon/api.go

```diff
@@ -1,25 +1,51 @@
 package beacon
 
-import "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/api"
+import (
+	"github.com/tendermint/tendermint/abci/types"
+
+	beacon "github.com/oasisprotocol/oasis-core/go/beacon/api"
+	"github.com/oasisprotocol/oasis-core/go/consensus/api/transaction"
+	"github.com/oasisprotocol/oasis-core/go/consensus/tendermint/api"
+	beaconState "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/apps/beacon/state"
+	genesis "github.com/oasisprotocol/oasis-core/go/genesis/api"
+)
 
 const (
 	// AppID is the unique application identifier.
-	AppID uint8 = 0x04
+	AppID uint8 = 0x08
 
 	// AppName is the ABCI application name.
-	// Run before the scheduler application.
-	AppName string = "100_beacon"
+	// Run before all other applications.
+	AppName string = "000_beacon"
 )
 
 var (
-	// EventType is the ABCI event type for beacon events.
+	// EventType is the ABCI event type for beacon/epoch events.
 	EventType = api.EventTypeForApp(AppName)
 
-	// QueryApp is a query for filtering events processed by the
+	// QueryApp is the query for filtering events procecessed by the
 	// beacon application.
 	QueryApp = api.QueryForApp(AppName)
 
-	// KeyGenerated is the ABCI event attribute key for the new
-	// beacons (value is a CBOR serialized beacon.GenerateEvent).
-	KeyGenerated = []byte("generated")
+	// KeyEpoch is the ABCI event attribute for specifying the set epoch.
+	KeyEpoch = []byte("epoch")
+
+	// KeyBeacon is the ABCI event attribute key for the new beacons.
+	KeyBeacon = []byte("beacon")
+
+	// MethodSetEpoch is the method name for setting epochs.
+	MethodSetEpoch = transaction.NewMethodName(AppName, "SetEpoch", beacon.EpochTime(0))
+
+	// Methods is a list of all methods supported by the beacon application.
+	Methods = []transaction.MethodName{
+		MethodSetEpoch,
+		MethodSCRAPECommit,
+		MethodSCRAPEReveal,
+	}
```

Comment:
The reason why `MethodSetEpoch` was previously defined here (in an internal Tendermint-specific package) is because it was only used by mock epochtime. For non-mock methods it may be useful to actually define them in the beacon API package (e.g., `go/beacon/api/scrape.go`) and add methods for generating transactions.

That is unless the SCRAPE-based implementation is supposed to be an internal detail of the Tendermint consensus backend.

Reply:
I have no strong opinions of moving this or not.  It is deeply tied to the tendermint implementation for now, so I am ok with leaving things here.  When we get another consensus implementation (if ever), moving it is trivial as well.

Reply:
So I would like to avoid imports of consensus-backend internal stuff outside `go/consensus/tendermint`. Currently the beacon worker (in `go/worker/beacon`) needs these so maybe it would make sense to move them out to `go/beacon/api/scrape.go`. As you said it should be trivial and I think it makes the most sense. The "mock set epoch" method can stay here as it is not needed anywhere else.

Reply:
(The other option would be moving the beacon worker into the Tendermint consensus backend and not exposing any of the SCRAPE stuff in `go/beacon/api/scrape.go`.)

Reply:
I sort of think that we're fooling ourselves if we think that we are actually going to support a different consensus backend, and that the code will be a lot simpler if we assume we are not  going to do that ever, but ok.